File: 153-FindMinimuminRotatedSortedArray/153-FindMinimuminRotatedSortedArray.java
class Solution {
    public int findMin(int[] nums) {
        if (nums.length==1) return nums[0];
        int l = 0, r = nums.length - 1; // l = 0, r = 3
        while (l<r) {
            int midindex = l + (r-l) / 2; //midindex = 1 //midval = 14; =15
            if (nums[midindex] > nums[r]) { //nums[1] = 13 < 14 = midval ; [2]<;
                l = midindex + 1; //l = 1
            } else {
                r = midindex;
            }
        }
        return nums[l];
    }
}

/*
    1st pass 
    l=0
    r=1
    midindex=0
    midval = 1
    nums[0] = 2 > 1 => l = 1

    2nd pass
    l=0
    r=1
    midindex=0
    midval = nums[0] + nums[1] = 12 
    nums[0] = 11 < 12 => r=0

    nums[]
*/

File: 153-FindMinimuminRotatedSortedArray/153-FindMinimuminRotatedSortedArray.py
class Solution:
    def findMin(self, nums: List[int]) -> int:
        '''
        restraint: 
        -5000 <= element <= 5000
        no repeating
        0<nums.length
        no same increment

        input: array=[3, 4, 5, 1, 2] rotated array? => how many rotations
        output: 1

        0 1 2 3 4 5 6 index

        3 4 5 6 7 1 2

        (1+6)//2 = 3 mid index=> nums[index] = mid value => 6
        (3+2)//2 = 2 actual value

        (1+6)//2 = 3 mid index=> nums[index] = mid value => 6
        (3+2)//2 = 2 actual value

        (5+6)//2 = 5 mid index => nums[5] = 1
        (1+2)//2 = 5 actual value       

        idea:
        shift back to ascending order => move each element reference => time, space
        sorted => binary search. requires sorted array, if array is not then the search cant be done properly
        identifying the broken part

        neeed:
        we need l, r
        if mid > value then move r = mid
        ese l = mid + 1

        know where the pointers are going =>


        pseudo: 
        left, right pointer, mid

        while loop:
            mid = (left+right)//2 = left + (right-left)//2
            if mid value > actual value 
                left index = mid index + 1         
            if else
                right index = mid

        return nums[leftindex]
        '''
        #implement:
        #left, right pointer, mid
        leftIndex = 0
        rightIndex = len(nums) - 1
    
        while (leftIndex<rightIndex) :
            midIndex = leftIndex + (rightIndex-leftIndex)//2
            #if mid value > right value 
            if (nums[midIndex]>nums[rightIndex]):
                leftIndex = midIndex + 1
            else:
                rightIndex = midIndex
        return nums[leftIndex]